squad/ci/management/commands/listen.py

```diff
@@ -0,0 +1,110 @@
+import logging
+import multiprocessing
+import signal
+import sys
+import time
+from django.core.management.base import BaseCommand
+from django.db.models import Field
+
+
+from squad.ci.models import Backend
+
+
+logger = logging.getLogger()
+
+
+class Listener(multiprocessing.Process):
+
+    def __init__(self, backend):
+        self.backend = backend
+        self.implementation = backend.get_implementation()
+        super(Listener, self).__init__()
+
+    def run(self):
+        backend = self.backend
+        impl = self.implementation
+
+        logger.info("Backend %s starting" % backend.name)
+        signal.signal(signal.SIGINT, self.stop)
+        signal.signal(signal.SIGTERM, self.stop)
+        impl.listen()
+        logger.info("Backend %s exited on its own" % backend.name)
+
+    def stop(self, signal, stack_frame):
+        logger.info("Backend %s finishing ..." % self.backend.name)
+        sys.exit()
+
+
+class ListenerManager(object):
+
+    def __init__(self):
+        self.__processes__ = {}
+
+    def run(self):
+        self.setup_signals()
+        self.loop()
+        self.cleanup()
+
+    def setup_signals(self):
+        # make SIGTERM equivalent to SIGINT (e.g. control-c)
+        signal.signal(signal.SIGTERM, signal.getsignal(signal.SIGINT))
+
+    def keep_listeners_running(self):
+        ids = list(self.__processes__.keys())
+
+        for backend in Backend.objects.all():
+            process = self.__processes__.get(backend.id)
+            if process:
+                # already running: restart if needed
+                if fields(backend) != fields(process.backend):
+                    self.stop(backend.id)
+                    self.start(backend)
+            else:
+                # not running, just start
+                self.start(backend)
+            if backend.id in ids:
```

Comment:
there is no race condition: `ids` is initialized as a _copy_ of `self.__processes__.keys()` so they are different lists.